src/data/fintype/order.lean

```diff
@@ -0,0 +1,91 @@
+/-
+Copyright (c) 2021 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+
+import order.lattice
+import data.fintype.basic
+
+/-!
+# Order structures on finite sets
+
+This file shows that a `nonempty` `lattice` on a `fintype` gives a `complete_lattice`, and that a
+`linear_order` on a nonempty `fintype` gives a `complete_linear_order`. The latter applies in
+particular to `fin (n+1)`. Getting to a `bounded_lattice` from a `lattice` is computable, but the
+subsequent definitions are not, since the definitions of `Sup` and `Inf` use `set.to_finset`, which
+implicitly requires a `decidable_pred` instance for every `s : set α`.
+
+An explicit instance is given for a `complete_linear_order` on `fin (n+1)`, but the rest are given
+as `def`s, to avoid loops in instance searches.
+-/
+
+section main
+
+/-- a finite inhabited `semilattice_inf` has a `⊥` -/
+def fintype.semilattice_inf_bot (α : Type*) [inhabited α] [fintype α] [semilattice_inf α] :
+  semilattice_inf_bot α :=
+let bot := finset.fold (⊓) (arbitrary α) id finset.univ in
+{ bot := bot,
+  bot_le := λ a, ((@finset.fold_op_rel_iff_and _ _ (⊓) _ _ _ _ _ _
+    (λ _ _ _, le_inf_iff) bot).mp le_rfl).2 a (finset.mem_univ _),
+  .. (infer_instance : semilattice_inf α)}
+
+/-- a finite inhabited `semilattice_sup` has a `⊤` -/
+def fintype.semilattice_sup_top (α : Type*) [inhabited α] [fintype α] [semilattice_sup α] :
+  semilattice_sup_top α :=
+let top := finset.fold (⊔) (arbitrary α) id finset.univ in
+{ top := top,
+  le_top := λ a, (((@finset.fold_op_rel_iff_and _ _ (⊔) _ _ _ (arbitrary α) _ (λ x y, y ≤ x)
+  (λ _ _ _, sup_le_iff) top )).mp le_rfl).2 a (finset.mem_univ a),
+  .. (infer_instance : semilattice_sup α)}
```

Comment:
I think we can show the existence of `inf` on a fintype given `semilattice_sup_top`, right? Perhaps these constructions belong in this file; though I'm happy for them to be later PRs rather than making this one any bigger.

Reply:
You're right, my mistake.